File: ios/WolManager/Util/RemoteDesktop.swift
import UIKit

/*
 * Remotedesktop via Microsoft Remote Desktop / Windows App.
 * iOS kann keinen eigenen RDP-Client bereitstellen → wir übergeben die Verbindung
 * an die installierte App. Die Windows App legt beim Aufruf ein Profil an, das auf
 * dem Gerät bestehen bleibt (gewollt – im Gegensatz zur Windows-Desktop-Version der
 * App wird hier nichts wieder gelöscht).
 *
 * ── Was die Windows App (mobil) laut Microsoft-Doku kann und was nicht ──
 * Legacy-Schema `rdp://<attribut>=<typ>:<wert>&…` (gilt für macOS/iOS/Android):
 *   • `full address=s:<host>` und `username=s:<user>` werden übernommen.
 *   • Es gibt KEIN Passwort-Attribut und KEIN Friendly-/Anzeigename-Attribut.
 * `ms-rd://` ist offiziell NUR für den Windows Desktop Client (MSRDC) → auf iOS
 *   meist ohne Handler, bleibt aber als letzter Kandidat erhalten.
 *
 * ⇒ Passwort: einziger dokumentierter Weg = Zwischenablage (UI zeigt Hinweis-Toast).
 * ⇒ Anzeigename = Gerätename: nur über eine übergebene `.rdp`-Datei erreichbar
 *   (die Windows App benennt das neue Profil nach dem Dateinamen). Da iOS eine
 *   Datei nicht programmatisch in einer fremden App öffnen kann, führt die Bridge
 *   dazu das Freigabe-Sheet auf. Diese Datei enthält bewusst KEIN Passwort (mobile
 *   Clients lesen `password:` ohnehin nicht, und Klartext im Teilungspfad wäre ein
 *   unnötiges Risiko).
 *
 * ── Wichtiger Parsing-Fix ──
 * Seit iOS 17 gilt in `URL(string:)` striktes RFC-3986-Parsing: ein rohes `=`/`:`
 * im Autoritäts-Teil (z. B. `full%20address=s:host`) macht die URL `nil` → der
 * Kandidat wurde bislang still übersprungen („nicht installiert", obwohl die
 * Windows App vorhanden war). Deshalb werden die Attribut-Trenner jetzt selbst
 * prozentkodiert (`%3D`/`%3A`/`%26`); die Windows App dekodiert sie vor dem
 * Auswerten der Attribute (genau so wie sie bereits `full%20address` dekodiert).
 */
enum RemoteDesktop {
    static let errNotInstalled = "remote.notinstalled"
    static let errNoHost = "remote.nohost"

    struct Result {
        let host: String
        let username: String
        let passwordCopied: Bool
        let hasPassword: Bool
        /// true, wenn kein URI-Handler griff und stattdessen eine `.rdp`-Datei
        /// erzeugt wurde (Bridge zeigt dann das Freigabe-Sheet).
        let viaFile: Bool
        let fileUrl: URL?
        let fileName: String
    }

    enum RemoteError: Error, Equatable {
        case noHost
        case notInstalled
    }

    /// Host für die RDP-Verbindung: IP/Hostname, sonst Gerätename.
    static func hostOf(ip: String, name: String) -> String {
        let t = ip.trimmingCharacters(in: .whitespacesAndNewlines)
        return t.isEmpty ? name.trimmingCharacters(in: .whitespacesAndNewlines) : t
    }

    /// URI-Kandidaten in Reihenfolge: legacy `rdp://` (dokumentiert für iOS, mit
    /// `full address` + `username`), dann `ms-rd://add/host/` (MSRDC-Form).
    /// Alle Attribut-Trenner werden prozentkodiert, damit `URL(string:)` (iOS 17+
    /// strikt nach RFC 3986) die URI überhaupt akzeptiert.
    static func candidates(host: String, username: String, mode: String = "full") -> [String] {
        let h = encodeValue(host.trimmingCharacters(in: .whitespacesAndNewlines))
        guard !h.isEmpty else { return [] }
        let u = encodeValue(username.trimmingCharacters(in: .whitespacesAndNewlines))
        // Legacy: `full address=s:<host>` → `full%20address%3Ds%3A<host>`; der Wert
        // behält sein eigenes (bereits kodiertes) Escaping.
        var legacy = "rdp://full%20address%3Ds%3A" + h
        if !u.isEmpty { legacy += "%26username%3Ds%3A" + u }
        var query: [String] = ["use.maximizewindow=\(mode == "full" ? "true" : "false")"]
        if !u.isEmpty { query.insert("username=\(u)", at: 0) }
        let msrd = "ms-rd://add/host/\(h)?" + query.joined(separator: "&")
        return [legacy, msrd]
    }

    /// Wert-Escaping für URI-Attribute: wie Android (RFC 3986, ohne Sub-delims).
    /// `:` ist hier NICHT erlaubt, obwohl Hostnamen es enthalten dürfen (`host:3389`):
    /// Die Legacy-URI liegt ohne `?` im Autoritäts-Teil, und iOS 17+ verwirft bei
    /// rohen Doppelpunkten die gesamte URL. Prozentkodiert dekodiert die Windows App
    /// den Wert korrekt zurück.
    static func encodeValue(_ value: String) -> String {
        let allowed = CharacterSet(charactersIn:
            "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~@/")
        return value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
    }

    /// Inhalt einer `.rdp`-Datei (KEY:TYP:WERT, CRLF) – Format wie die Desktop-App
    /// (`wol_app/utils.py`), aber ohne Passwort: mobile Clients lesen `password:`
    /// nicht, und der Anzeigename kommt aus dem Dateinamen.
    static func buildContent(host: String, username: String, mode: String) -> String {
        var lines = [
            "full address:s:\(host.trimmingCharacters(in: .whitespacesAndNewlines))",
            // Selbstsignierte Zertifikate (typisch für xrdp/Linux) ohne Rückfrage akzeptieren.
            "authentication level:i:0",
            // Adresse nach Redirection-Hop als Serveridentität behalten (xrdp).
            "use redirection server name:i:1",
        ]
        let u = username.trimmingCharacters(in: .whitespacesAndNewlines)
        if !u.isEmpty { lines.append("username:s:\(u)") }
        lines.append("screen mode id:i:\(mode == "full" ? 1 : 2)")
        return lines.joined(separator: "\r\n") + "\r\n"
    }

    /// Dateiname (ohne Endung) aus dem Gerätenamen: dateisystemsicher; die Windows
    /// App verwendet ihn als Anzeigename des neuen Profils.
    static func sanitizedFilename(_ name: String) -> String {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        let allowed = CharacterSet(charactersIn:
            "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789 -._")
        var out = String(trimmed.unicodeScalars.filter { allowed.contains($0) })
            .trimmingCharacters(in: .whitespaces)
        if out.isEmpty { out = "Remote-PC" }
        if out.count > 60 { out = String(out.prefix(60)).trimmingCharacters(in: .whitespaces) }
        return out
    }

    /// `.rdp`-Datei in den Cache schreiben (bleibt, bis iOS den Räumungsdienst aufruft).
    static func writeRdpFile(name: String, host: String, username: String, mode: String) throws -> URL {
        let dir = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("rdp", isDirectory: true)
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let url = dir.appendingPathComponent("\(name).rdp")
        let data = buildContent(host: host, username: username, mode: mode).data(using: .utf8) ?? Data()
        try data.write(to: url, options: .atomic)
        return url
    }

    /// Öffnet die Windows App für das Gerät; Passwort → Zwischenablage.
    /// Wichtig: die Zwischenablage wird VOR dem Öffnen gefüllt – ab iOS 16
    /// ist der Pasteboard-Zugriff aus dem Hintergrund eingeschränkt, und die
    /// App wechselt mit `open(url)` sofort in den Hintergrund. Vorher wird
    /// per `canOpenURL` geprüft, damit bei fehlender App nichts in die
    /// Zwischenablage gelangt.
    ///
    /// Greift kein URI-Handler, wird eine `.rdp`-Datei erzeugt und mit
    /// `viaFile` zurückgegeben; die Bridge übergibt sie dann per Freigabe-Sheet
    /// an die Windows App (Profilname = Gerätename). Nur wenn auch das nicht
    /// möglich ist (Dateiwrite schlug fehl), meldet die Bridge
    /// "remote.notinstalled"; fehlender Host → "remote.nohost".
    @discardableResult
    static func open(device: Device, password: String, mode: String) async throws -> Result {
        let host = hostOf(ip: device.ip, name: device.name)
        guard !host.isEmpty else { throw RemoteError.noHost }
        let uri = candidates(host: host, username: device.username, mode: mode)
        let fileName = sanitizedFilename(device.name)
        let launched = await MainActor.run { () -> Bool in
            for u in uri {
                guard let url = URL(string: u), UIApplication.shared.canOpenURL(url) else { continue }
                if !password.isEmpty { UIPasteboard.general.string = password }
                UIApplication.shared.open(url, options: [:], completionHandler: nil)
                return true
            }
            return false
        }
        if !launched {
            // Kein URI-Handler → `.rdp`-Datei für die Datei-Übergabe vorhalten.
            let url = try? writeRdpFile(name: fileName, host: host,
                                        username: device.username, mode: mode)
            if url == nil { throw RemoteError.notInstalled }
            return Result(host: host, username: device.username,
                          passwordCopied: false, hasPassword: !password.isEmpty,
                          viaFile: true, fileUrl: url, fileName: fileName)
        }
        return Result(host: host, username: device.username,
                      passwordCopied: !password.isEmpty, hasPassword: !password.isEmpty,
                      viaFile: false, fileUrl: nil, fileName: fileName)
    }
}

File: android_html/app/src/main/java/de/wolmanager/html/util/RemoteDesktop.kt
package de.wolmanager.html.util

/**
 * Remote-Desktop-Aufruf über die installierte **Windows App**
 * (früher „Microsoft Remote Desktop", `com.microsoft.rdc.androidx`).
 *
 * Android kann kein eigenes RDP bereitstellen, und die Windows App kennt — anders
 * als `mstsc` unter Windows — keinen Weg, einem Aufruf ein **Passwort** mitzugeben:
 * das dokumentierte URI-Schema der Client-App für macOS/iOS/Android
 * (`rdp://<attribut>=<typ>:<wert>&…`) kennt für `full address` und `username`
 * Attribute, aber keines für Passwörter (das Passwort könnte ohnehin nur der
 * Windows Credential Manager übernehmen, der Drittanbietern nicht zur Verfügung
 * steht). Die UI legt das Passwort deshalb in die Zwischenablage, damit es im
 * Verbindungsfenster eingefügt werden kann.
 *
 * Ebenso gibt es kein Attribut für den Anzeigenamen: der Eintrag in der Windows App
 * heißt wie die Adresse. Der Gerätename wird daher nur protokolliert.
 *
 * Quelle: Microsoft Learn – „Remote Desktop URI scheme" (Legacy `rdp://`-Schema,
 * Attribute `full address=s:`, `username=s:`; für Android dokumentiert).
 *
 * Rein Kotlin (kein `android.net.Uri`), damit es in JVM-Unit-Tests prüfbar ist.
 */
object RemoteDesktop {

    /** Fehler-Schlüssel: keine Windows App / kein RDP-Client installiert. */
    const val ERR_NOT_INSTALLED = "remote.notinstalled"

    /** Fehler-Schlüssel: für das Gerät ist weder IP/Hostname noch Name hinterlegt. */
    const val ERR_NO_HOST = "remote.nohost"

    /** Host-Kandidat: IP-Adresse, sonst der Gerätename (wie iOS). */
    fun hostOf(ip: String, name: String): String =
        ip.trim().ifBlank { name.trim() }

    /**
     * URI-Kandidaten in Reihenfolge ihrer Wahrscheinlichkeit. `rdp://` ist das für
     * Android dokumentierte Schema; `ms-rd://add/host/…` (iOS-Form) als Fallback für
     * App-Versionen, die zusätzlich dieses Schema registrieren.
     *
     * [mode] (`"full"` / `"win"`) wird akzeptiert, ist aber ohne Wirkung: die
     * Attributtabelle des Android-Clients enthält kein `screen mode id` (nur macOS/iOS).
     * Auf dem Smartphone ist die Sitzung ohnehin randlos.
     */
    fun candidates(host: String, username: String, mode: String = "full"): List<String> {
        val h = encodeValue(host.trim())
        if (h.isEmpty()) return emptyList()
        val u = encodeValue(username.trim())
        val attrs = buildString {
            append("full%20address=s:").append(h)
            if (u.isNotEmpty()) append("&username=s:").append(u)
        }
        return listOf(
            "rdp://$attrs",
            "ms-rd://add/host/$h" + if (u.isNotEmpty()) "?username=$u" else "",
        )
    }

    /**
     * Percent-Kodierung für Werte im Query-Teil: erhalten bleibt, was ein Hostname
     * oder Benutzername harmlos verwendet (`A-Za-z0-9`, `-._~`, `:`, `@`), alles
     * andere — insbesondere `%`, `&`, `=`, `#`, `?`, `\`, Leerzeichen und
     * Nicht-ASCII — wird als UTF-8-Einzelbyte `%XX` geschrieben.
     */
    fun encodeValue(value: String): String {
        if (value.isEmpty()) return ""
        val sb = StringBuilder(value.length + 8)
        for (ch in value) {
            when {
                ch in 'A'..'Z' || ch in 'a'..'z' || ch in '0'..'9' -> sb.append(ch)
                ch in "-._~:@/" -> sb.append(ch)
                else -> for (b in ch.toString().toByteArray(Charsets.UTF_8)) {
                    sb.append('%').append(HEX[(b.toInt() shr 4) and 0x0F]).append(HEX[b.toInt() and 0x0F])
                }
            }
        }
        return sb.toString()
    }

    /**
     * Inhalt einer `.rdp`-Datei (`KEY:TYP:WERT`, CRLF) — Format wie die Desktop-App
     * (`wol_app/utils.py`). Mit Passwort: `password:54:` = base64(UTF-16LE), wie von
     * mstsc dokumentiert. Die mobilen Clients lesen es möglicherweise nicht — dann
     * bleibt der Zwischenablage-Weg der Brücke. Anders als iOS (Freigabe-Sheet!)
     * bleibt die Datei hier im app-private Cache und wird nur der Windows App mit
     * zeitlich begrenzter Leserechte übergeben, deshalb ist das Einbetten vertretbar.
     *
     * Der Anzeigename des Profils ergibt sich aus dem **Dateinamen** (siehe
     * [sanitizedFilename]) — die `.rdp`-Spezifikation selbst kennt kein Namensfeld.
     */
    fun buildRdpContent(host: String, username: String, password: String, mode: String): String {
        val lines = mutableListOf(
            "full address:s:" + host.trim(),
            // Selbstsignierte Zertifikate (typisch für xrdp/Linux) ohne Rückfrage akzeptieren.
            "authentication level:i:0",
            // Adresse nach Redirection-Hop als Serveridentität behalten (xrdp).
            "use redirection server name:i:1",
        )
        val u = username.trim()
        if (u.isNotEmpty()) lines.add("username:s:$u")
        if (password.isNotEmpty()) {
            // java.util.Base64: ab API 26 (minSdk) verfügbar und JVM-testbar.
            val b64 = java.util.Base64.getEncoder()
                .encodeToString(password.toByteArray(Charsets.UTF_16LE))
            lines.add("password:54:$b64")
            lines.add("prompt credential:i:0")
        }
        lines.add("screen mode id:i:" + if (mode == "full") 1 else 2)
        return lines.joinToString("\r\n") + "\r\n"
    }

    /**
     * Dateiname (ohne Endung) aus dem Gerätenamen: dateisystemsicher machen. Die
     * Windows App verwendet ihn als Anzeigename der neu angelegten Verbindung.
     */
    fun sanitizedFilename(name: String): String {
        var out = name.trim().map { if (it in ('A'..'Z') || it in ('a'..'z') || it in ('0'..'9') ||
            it in " -._") it else ' ' }.joinToString("").trim()
        if (out.isEmpty()) out = "Remote-PC"
        if (out.length > 60) out = out.substring(0, 60).trim()
        return out
    }

    /** Paketname der Windows App (ehem. „Microsoft Remote Desktop", Android). */
    const val WINDOWS_APP_PACKAGE = "com.microsoft.rdc.androidx"

    private const val HEX = "0123456789ABCDEF"
}
